### PR TITLE
Add scaleStrategy.maxUnavailable for CloneSet

### LINKED ISSUE #761: [feature request] support maxUnavailable in scaleStrategy for CloneSet
**What would you like to be added**:
 add a maxUnavailable field for scale strategy for CloneSet

**Why is this needed**:
Some backend service need to control the changes rate of replicas so as to minimize the impact of the upstream service. For example, a maximum N replicas should be added in 5 minutes.   The maxUnavailable in scaleStrategy can enable such limitation. 

```
apiVersion: apps.kruise.io/v1alpha1
kind: CloneSet
metadata:
  name: sample-cloneset
  minReadySeconds: 60 
  scaleStrategy:
    maxUnavailable: 1  
```

In the above example, one can ensure maximum 1 replicas is added every minutes

### PR DESCRIPTION
**Pull request by veophi**
Signed-off-by: veophi <vec.g.sun@gmail.com>

### Ⅰ. Describe what this PR does
Add scaleStrategy.maxUnavailable for CloneSet, and this field works only when expanding.

### Ⅱ. Does this pull request fix one issue?

#fixes #761

### Ⅲ. Describe how to verify it
e2e test

### CODE REVIEW

File: apis/apps/v1alpha1/cloneset_types.go [hunk 1]
```diff
@@ -80,6 +80,12 @@ type CloneSetScaleStrategy struct {
 	// PodsToDelete is the names of Pod should be deleted.
 	// Note that this list will be truncated for non-existing pod names.
 	PodsToDelete []string `json:"podsToDelete,omitempty"`
+	// The maximum number of pods that can be unavailable for scaled pods.
+	// This field can control the changes rate of replicas for CloneSet so as to minimize the impact for users' service.
+	// The scale will fail if the number of unavailable pods were greater than this MaxUnavailable at scaling up.
+	// MaxUnavailable works only when scaling up.
+	// Default to math.MaxInt32
```
Comment (FillZpp) [inline apis/apps/v1alpha1/cloneset_types.go:87]: Defaults to nil

File: pkg/controller/cloneset/sync/cloneset_sync_utils.go [hunk 2]
```diff
@@ -67,10 +69,10 @@ func (e expectationDiffs) isEmpty() bool {
 
 // This is the most important algorithm in cloneset-controller.
 // It calculates the pod numbers to scaling and updating for current CloneSet.
-func calculateDiffsWithExpectation(cs *appsv1alpha1.CloneSet, pods []*v1.Pod, currentRevision, updateRevision string) (res expectationDiffs) {
+func calculateDiffsWithExpectation(cs *appsv1alpha1.CloneSet, pods []*v1.Pod, currentRevision, updateRevision string, recorder record.EventRecorder) (res expectationDiffs) {
```
Comment (FillZpp) [inline pkg/controller/cloneset/sync/cloneset_sync_utils.go:72]: Do not pass `recorder` into the calculation function. It should be recorded outside.

File: pkg/controller/cloneset/sync/cloneset_sync_utils.go [hunk 3]
```diff
@@ -59,6 +60,8 @@ type expectationDiffs struct {
 	updateNum int
 	// updateMaxUnavailable is the maximum number of ready Pods that can be updating
 	updateMaxUnavailable int
+	// scaleUpLimit limits the maximum number of creating pods when scaling up
+	scaleUpLimit int
```
Comment (FillZpp) [inline pkg/controller/cloneset/sync/cloneset_sync_utils.go:64]: Put `scaleUpLimit` along with `deleteReadyLimit`.

File: pkg/controller/cloneset/sync/cloneset_scale.go [hunk 4]
```diff
@@ -70,10 +71,14 @@ func (r *realControl) Scale(
 	diffRes := calculateDiffsWithExpectation(updateCS, pods, currentRevision, updateRevision)
 	updatedPods, notUpdatedPods := clonesetutils.SplitPodsByRevision(pods, updateRevision)
 
+	if diffRes.scaleNum > 0 && diffRes.scaleUpLimit <= 0 {
```
Comment (FillZpp) [inline pkg/controller/cloneset/sync/cloneset_scale.go:74]: Shoud it be:
`if diffRes.scaleNum > 0 && diffRes.scaleUpLimit < diffRes.scaleNum {`

Comment (veophi) [reply pkg/controller/cloneset/sync/cloneset_scale.go]: done

File: pkg/controller/cloneset/sync/cloneset_sync_utils.go (hunk 3 above)
Comment (veophi) [reply pkg/controller/cloneset/sync/cloneset_sync_utils.go]: done

**Pull request merged**